compiler/src/iree/compiler/Codegen/SPIRV/SPIRVEraseStorageBufferStaticShape.cpp

```diff
@@ -0,0 +1,184 @@
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-spirv-erase-storage-buffer-static-shape"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+class EraseStorageBufferStaticShapePass final
+    : public SPIRVEraseStorageBufferStaticShapeBase<
+          EraseStorageBufferStaticShapePass> {
+  void runOnOperation() override;
+};
+
+/// Returns true if the given `subspanOp` is from a 1-D static shaped storage
+/// buffer.
+bool is1DStaticShapedStorageBuffer(
+    IREE::HAL::InterfaceBindingSubspanOp subspanOp) {
+  auto type = subspanOp.getType().dyn_cast<MemRefType>();
+  if (!type) return false;
+  auto attr =
+      type.getMemorySpace().dyn_cast_or_null<IREE::HAL::DescriptorTypeAttr>();
+  if (!attr) return false;
+  return type.hasStaticShape() && type.getRank() == 1 &&
+         attr.getValue() == IREE::HAL::DescriptorType::StorageBuffer;
+}
+
+/// Rewrites a subspan op with 1-D static shape into dynamic shape.
+/// e.g.,
+///
+/// ```mlir
+///  hal.interface.binding.subspan set(0) binding(0) offset(%offset)
+///      : memref<16xf32>
+/// ```
+///
+/// is re-written to
+///
+/// ```mlir
+///  hal.interface.binding.subspan set(0) binding(0) offset(%offset)
+///      : memref<?xf32>{%c16}
+/// ```
+IREE::HAL::InterfaceBindingSubspanOp rewriteStorageBufferSubspanOp(
+    RewriterBase &rewriter, IREE::HAL::InterfaceBindingSubspanOp subspanOp) {
+  assert(is1DStaticShapedStorageBuffer(subspanOp));
+  LLVM_DEBUG({
+    llvm::dbgs() << "Rewriting subspan op: ";
+    subspanOp->print(llvm::dbgs(), OpPrintingFlags().assumeVerified());
+    llvm::dbgs() << "\n";
+  });
+
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(subspanOp);
+
+  auto oldType = subspanOp.getType().cast<MemRefType>();
+  auto newType =
+      MemRefType::get({ShapedType::kDynamic}, oldType.getElementType(),
+                      oldType.getLayout(), oldType.getMemorySpace());
+
+  SmallVector<Value, 1> dynamicDims;
+  assert(subspanOp.getDynamicDims().empty());
+  dynamicDims.push_back(rewriter.create<arith::ConstantIndexOp>(
+      subspanOp.getLoc(), oldType.getNumElements()));
+
+  auto newOp = rewriter.create<IREE::HAL::InterfaceBindingSubspanOp>(
+      subspanOp.getLoc(), newType, subspanOp.getSetAttr(),
+      subspanOp.getBindingAttr(), subspanOp.getDescriptorTypeAttr(),
+      subspanOp.getByteOffset(), dynamicDims, subspanOp.getAlignmentAttr(),
+      subspanOp.getDescriptorFlagsAttr());
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Rewritten to: ";
+    newOp->print(llvm::dbgs(), OpPrintingFlags().assumeVerified());
+    llvm::dbgs() << "\n";
+  });
+  return newOp;
+}
+
+/// Replace `origValue` with `replacementValue`. The replacement might
+/// require replacing the users of `origValue`. The results of the users
+/// themselves have to be replaced with results of the new users.
+void replaceUsesTransitively(RewriterBase &rewriter, Location loc,
```

Comment:
Did you want to duplicate this for some reason? I can move the same method to a `Common/Utils.h` after I submit my changes (actually going to replace a similar function that already existed).

Reply:
No; would be good to unify it!

Reply:
Ok, ill do it as a follow up here.